### DISCUSSION TITLE
Where is the documentation on how to perform CRUD using Mapped Class syntax?

### DISCUSSION THREAD

**Discussion by zhao-li** [sqlalchemy/sqlalchemy#8784] [Usage Questions] [answered]
Hi Folks 👋

I am learning to use SqlAlchemy and am having some difficulty finding more information on how to use the Mapped Class API.

Is there a page that succinctly goes over the basic CRUD operations using the Mapped Class API?

I can find the documentation on using the `session` and `Query` APIs:
- https://docs.sqlalchemy.org/en/14/orm/session_basics.html#adding-new-or-existing-items
- https://docs.sqlalchemy.org/en/14/orm/query.html

but am drawing a blank when looking for basic usage for the Mapped Class API.

Sorry, if I am using the wrong terminology.  Maybe the below code examples can better illustrate what I am trying to say:

Where can I find documentation on this syntax:
```
Greeting.query.filter_by(id=greeting_id).delete()
Greeting.query.all()
Greeting.query.get(greeting_id)
greeting.message = 'updated attribute' # is there a way to do this with multiple attributes at once (described in the documentation)?
db_session.commit() # is there a way to avoid this extra statement?
```
instead of this syntax:
```
greeting = Greeting.query.get(greeting_id) # I'm okay with this syntax
db_session.add(greeting) # can't figure out how to do this using the Greeting class without the `session` class
setattr(greeting, attribute, 'updated attribute') # is there a way to do this with multiple attributes at once (described in the documentation)?
db_session.delete(greeting)
db_session.commit() # this seems to be required by both sets of syntax
```

Is `[class].query` the same as the `Query` class described here?: https://docs.sqlalchemy.org/en/14/orm/query.html

Thank you for your time 🙏

**Comment by zzzeek** [comment, marked as answer]
> Hi Folks wave
> 
> I am learning to use SqlAlchemy and am having some difficulty finding more information on how to use the Mapped Class API.
> 
> Is there a page that succinctly goes over the basic CRUD operations using the Mapped Class API?
> 
> I can find the documentation on using the `session` and `Query` APIs:
> 
>     * https://docs.sqlalchemy.org/en/14/orm/session_basics.html#adding-new-or-existing-items
> 
>     * https://docs.sqlalchemy.org/en/14/orm/query.html
> 
> 
> but am drawing a blank when looking for basic usage for the Mapped Class API.

the usual ORM use cases are detailed in the tutorial which presents a complete picture:  https://docs.sqlalchemy.org/en/14/tutorial/orm_data_manipulation.html

> 
> Where can I find documentation on this syntax:
> 
> ```
> Greeting.query.filter_by(id=greeting_id).delete()
> Greeting.query.all()
> Greeting.query.get(greeting_id)

those above are queries.   The Query interface is now legacy, but the tutorial for that particular interface can be seen in the legacy 1.x tutorial at https://docs.sqlalchemy.org/en/14/orm/tutorial.html#querying 

for modern information on querying, the ORM query guide for 2.0 is at: https://docs.sqlalchemy.org/en/20/orm/queryguide/index.html  and the 1.4 version is at https://docs.sqlalchemy.org/en/14/orm/queryguide.html .  The 2.0 version is more in-depth.


> greeting.message = 'updated attribute' # is there a way to do this with multiple attributes at once (described in the documentation)?

if working with an instance of Greeting, then Python attribute set operations are how their values are modified.

if you are looking to just emit an UPDATE statement directly, for 1.4 see https://docs.sqlalchemy.org/en/14/orm/session_basics.html#update-and-delete-with-arbitrary-where-clause, again in 2.0 the organization / docs for all of this is vastly better laid out which you'd see in the queryguide at https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html .

But you want to definitely work with the ORM tutorial first.



> db_session.commit() # is there a way to avoid this extra statement?

the ORM always has an explicit transaction boundary.   you can use context managers with session.begin(), as described at the end of https://docs.sqlalchemy.org/en/14/orm/session_basics.html#framing-out-a-begin-commit-rollback-block .  but implicit transactions in SQLAlchemy are deprecated and they are a bad idea.


> ```
> ```
> greeting = Greeting.query.get(greeting_id) # I'm okay with this syntax
> db_session.add(greeting) # can't figure out how to do this using the Greeting class without the `session` class

the add() is unnecessary, greeting is already persistent

> setattr(greeting, attribute, 'updated attribute') # is there a way to do this with multiple attributes at once (described in the documentation)?

Python attribute assignment only works on one attribute at a time.  you can add a method to your Greeting class called "set_attributes()", like:

```py
def set_attributes(self, **kw):
    for k in kw:
       setattr(self, k, kw[k])
```



> db_session.commit() # this seems to be required by both sets of syntax

always

> ```
> 
> Is `[class].query` the same as the `Query` class described here?: https://docs.sqlalchemy.org/en/14/orm/query.html

``class.query`` is a non-standard syntax used by the Flask-SQLAlchemy library. That's not part of SQLAlchemy (but yes it's the same as session.query(class)).

**Reply by zhao-li** [reply]
Thank you so much for that thorough write up.  I'll take a closer look at the links and circle back.  Thanks again for your time 🙏

**Reply by zhao-li** [reply]
Thanks again @zzzeek for providing so much information.

I took a closer look at your write-up and have some follow up questions:

# CRUD
> the usual ORM use cases are detailed in the tutorial which presents a complete picture: https://docs.sqlalchemy.org/en/14/tutorial/orm_data_manipulation.html

Thank you for pointing out that this is the de facto tutorial to follow.

## Create
Is it safe to say that the only way to create a new instance/row is?:
```
greeting = Greeting('hello, world')
db_session.add(greeting)
db_session.commit()
```
There's nothing like this?:
```
greeting = Greeting('hello, world')
greeting.save() # something equivalent to this?
db_session.commit()
```

## Read
This is the only way to retrieve an instance/row?:
```
db_session.get(Greeting, greeting_id)
Greeting.query.get(greeting_id)
```
There's nothing like this?:
```
Greeting.get(greeting_id)
```

## Delete
This is the only way to delete an instance/row?:
```
greeting = db_session.get(Greeting, greeting_id)
db_session.delete(greeting)
db_session.commit()
or
Greeting.query.filter_by(id=greeting_id).delete()
db_session.commit()
```
There's nothing like this?:
```
greeting.delete()
```

I guess I just want to confirm this before giving up searching for it.

# Query Interface
Regarding the `Query` interface:
> The Query interface is now legacy

What is the new querying API/interface referred to as?

Thank you for linking the 2.0 and 1.4 documentation.  After reading through those links and figuring out how to navigate to/from each version's documentation, it really helped me understand the landscape much easier.

I'm not sure what I might be missing, but the older, deprecated `Query` interface seems to be more in line with the other ORMs I've used in other languages.  The `Query` interface from the 1.4 documentation seems to abstract away the SQL statements, whereas, the newer 2.0 documentation seems to be putting the developer back into SQL.

Is that right or am I missing something?

# Flask-SQLAlchemy
Thank you for pointing out Flask-SQLAlchemy to me.  I didn't know it existed.  Ironically, I was able to better understand their documentation a bit better.

I think this helps sums things up for me: https://flask-sqlalchemy.palletsprojects.com/en/3.0.x/queries/

They listed out the CUD of CRUD and then proceeded to explain that R (or querying) is transitioning from legacy `Query` (e.g. `Model.query` or `session.query`) is getting replaced by `session.execute(select(...))` at the very end.

I'm a bit bummed because it feels like `Model.query` abstracted away more of the SQL than `session.execute(select(...))`.  Am I thinking about this wrong?

I think I was looking for something similar to what I've been used to and that is causing me to continually search for something that might not exist.

Thanks again for all of your time and effort in helping me get up to speed on SQLAlchemy 🙏

**Reply by CaselIT** [reply]
> ## Create
> 
> There's nothing like this?:
> 
> ```
> greeting = Greeting('hello, world')
> greeting.save() # something equivalent to this?
> db_session.commit()
> ```

No

> ## Read
> There's nothing like this?:
> 
> ```
> Greeting.get(greeting_id)
> ```

No

> ## Delete
> There's nothing like this?:
> 
> ```
> greeting.delete()
> ```

No

> I guess I just want to confirm this before giving up searching for it.
> # Query Interface
> 
> Regarding the `Query` interface:
> 
> > The Query interface is now legacy
> 
> What is the new querying API/interface referred to as?
> 
> Thank you for linking the 2.0 and 1.4 documentation. After reading through those links and figuring out how to navigate to/from each version's documentation, it really helped me understand the landscape much easier.
> 
> I'm not sure what I might be missing, but the older, deprecated `Query` interface seems to be more in line with the other ORMs I've used in other languages. The `Query` interface from the 1.4 documentation seems to abstract away the SQL statements, whereas, the newer 2.0 documentation seems to be putting the developer back into SQL.

You seem to be looking for an [Active record](https://en.wikipedia.org/wiki/Active_record_pattern) orm. SQLAlchemy is a [Data Mapper](https://en.wikipedia.org/wiki/Data_mapper_pattern) orm.
If you prefer active record you should probably look into something like django orm that uses that pattern https://docs.djangoproject.com/en/4.1/topics/db/

**Reply by zhao-li** [reply]
@CaselIT ah yes, that's it.  Active Record vs Data Mapper.  That is what I've been missing.  Thank you for pointing that out and reminding me of it.

I can rest easy that I haven't missed any documentation or tutorial that goes through a more Active Record syntax.  Thank you for putting that nagging feeling to rest.

Ironically, I was working with Django (and its ORM) in the past but had to switch to Flask for this current project.

I Google'd a bit and found out there's another Active Record ORM called PeeWee that might work with Flask, since it doesn't seem like getting Django ORM to work with Flask is super easy.  Please let me know if you have tips on using an Active Record ORM with Flask.

Or maybe I'll take this opportunity to learn about the Data Mapper side of things.

Thanks again for all of your help 🙏

**Reply by CaselIT** [reply]
>  Please let me know if you have tips on using an Active Record ORM with Flask.

No, sorry. I've never used flask for anything more than a few simple things. 

Regarding active record orm, I don't have much experience with them, you seem to already know more than me :)

**Reply by zhao-li** [reply]
Thank you @CaselIT and @zzzeek for all of your help, tips, and advice.  Thank you also for maintaining such a crucial piece of software.  Take care 🙏
